### PR TITLE
Move viewportCallback out of resources and into separate util.

### PR DESCRIPTION
**Pull request by samouri**
**wip**
- Introduces new `viewport-observer` file with `observe()` and `unobserve()` so that we can track all elements with the same InOb (specifically for viewportCallback)

partial for https://github.com/ampproject/amphtml/issues/30620

### CODE REVIEW

File: extensions/amp-timeago/0.1/amp-timeago.js [hunk 1]
```diff
@@ -61,13 +65,28 @@ export class AmpTimeAgo extends AMP.BaseElement {
 
     this.setFuzzyTimestampValue_();
     this.element.appendChild(this.timeElement_);
+
+    this.viewportObserver_ = getViewportObserver((records) => {
```
Comment (samouri) [inline extensions/amp-timeago/0.1/amp-timeago.js:69]: note: would we prefer one inob per timeago component, or one to track all timeago components per win?

Comment (kristoferbaxter) [reply extensions/amp-timeago/0.1/amp-timeago.js:69]: This is a good general question across all components. I'd lean toward the default pattern of one inob per component type.

Comment (dvoytenko) [reply extensions/amp-timeago/0.1/amp-timeago.js]: I think one per-type might be easier for now. But I think in the long-term the best strategy is one per use-case, i.e. where conflicts between observers cannot happen. That being said, the timeago definitely seems like a separate use case to me. So in this particular case, per-type == per-use-case.

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 2]
```diff
@@ -1232,6 +1233,7 @@ export class AmpA4A extends AMP.BaseElement {
 
   /** @override */
   layoutCallback() {
+    observe(this.element, (inViewport) => this.viewportCallback_(inViewport));
```
Comment (dvoytenko) [inline extensions/amp-a4a/0.1/amp-a4a.js:1236]: For ads it might be important to have this callback prior to `layoutCallback` has fired. Can we confirm this directly? IMHO, it might be worth pushing ads into a separate PR due to a higher probability of rollback and/or more research needed.

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 3]
```diff
@@ -1143,8 +1146,14 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
     }
   }
 
+  /** @override */
+  layoutCallback() {
```
Comment (dvoytenko) [inline extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js:1150]: I feel like we might need `super.layoutCallback()` called here.

File: extensions/amp-anim/0.1/amp-anim.js [hunk 4]
```diff
@@ -84,6 +82,7 @@ export class AmpAnim extends AMP.BaseElement {
 
   /** @override */
   layoutCallback() {
+    observe(this.element, () => this.updateInViewport_());
```
Comment (dvoytenko) [inline extensions/amp-anim/0.1/amp-anim.js:85]: I might be paranoid, but this is slightly different than what we had before. To make it equivalent, you'd need to do something like this:

```
layoutCallback() {
  ...
  return this.loadPromise(img).then(() => observe(this.element, ...));
}
```

I.e. this way it's guaranteed that the inViewport will not arrive before loading has completed.

File: extensions/amp-apester-media/0.1/amp-apester-media.js [hunk 5]
```diff
@@ -151,11 +156,12 @@ class AmpApesterMedia extends AMP.BaseElement {
       renderer: true,
       tags: extractTags(this.getAmpDoc(), this.element),
     };
+    observe(this.element, (inViewport) => this.viewportCallback_(inViewport));
   }
 
   /** @override */
   firstLayoutCompleted() {
-    this.viewportCallback(this.isInViewport());
+    this.viewportCallback_(this.isInViewport());
```
Comment (dvoytenko) [inline extensions/amp-apester-media/0.1/amp-apester-media.js:164]: We should probably no longer rely on `this.isInViewport()`? Also, maybe we should just do the same thing here that you did for `amp-anim` (with the same caveat that I commented on).

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 6]
```diff
@@ -100,13 +100,6 @@ class AmpBrightcove extends AMP.BaseElement {
     return isLayoutSizeDefined(layout);
   }
 
-  /** @override */
-  viewportCallback(visible) {
-    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {
```
Comment (dvoytenko) [inline extensions/amp-brightcove/0.1/amp-brightcove.js:105]: This is blocked on the video PR, right? We need to make sure we don't merge in the wrong order.

File: extensions/amp-carousel/0.1/base-carousel.js [hunk 7]
```diff
@@ -38,6 +39,7 @@ export class BaseCarousel extends AMP.BaseElement {
 
   /** @override */
   buildCallback() {
+    observe(this.element, (inViewport) => this.viewportCallback_(inViewport));
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/base-carousel.js:42]: Let's drop from here. `layoutCallback` is enough.

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js [hunk 8]
```diff
@@ -145,11 +145,6 @@ class AmpDailymotion extends AMP.BaseElement {
     return isLayoutSizeDefined(layout);
   }
 
-  /** @override */
-  viewportCallback(visible) {
-    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {visible});
```
Comment (dvoytenko) [inline extensions/amp-dailymotion/0.1/amp-dailymotion.js:150]: Ditto, re: video pr

File: extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js [hunk 9]
```diff
@@ -101,15 +101,6 @@ export class AmpFlyingCarpet extends AMP.BaseElement {
     );
   }
 
-  /** @override */
-  viewportCallback(inViewport) {
-    Services.ownersForDoc(this.element).updateInViewport(
```
Comment (dvoytenko) [inline extensions/amp-fx-flying-carpet/0.1/amp-fx-flying-carpet.js:106]: Same not on ads. This looks really nice, but I'm a bit concerned with bundling ads with other things due to higher probability of rollbacks.

File: extensions/amp-timeago/0.1/amp-timeago.js [hunk 10]
```diff
@@ -63,13 +64,23 @@ export class AmpTimeAgo extends AMP.BaseElement {
     this.element.appendChild(this.timeElement_);
   }
 
-  /** @override */
-  viewportCallback(inViewport) {
+  /** @param {boolean} inViewport */
```
Comment (dvoytenko) [inline extensions/amp-timeago/0.1/amp-timeago.js:67]: `@private`

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: extensions/amp-apester-media/0.1/amp-apester-media.js (hunk 5 above)
Comment (samouri) [reply extensions/amp-apester-media/0.1/amp-apester-media.js:164]: definitely. `this.isInVieport_` needs to be removed since it is dependent on the same pathway in resources

File: extensions/amp-anim/0.1/amp-anim.js (hunk 4 above)
Comment (samouri) [reply extensions/amp-anim/0.1/amp-anim.js:85]: Good point. I missed that nuance, since this layoutCallback returns a promise.

File: extensions/amp-apester-media/0.1/amp-apester-media.js (hunk 5 above)
Comment (samouri) [reply extensions/amp-apester-media/0.1/amp-apester-media.js:164]: making this change now, and spinning out all "lower risk" components from ads

Comment (samouri) [reply extensions/amp-apester-media/0.1/amp-apester-media.js:164]: actually, looks like there is a specific path to account for when this component is made visible but hasn't completed layout yet: (inViewport &&  !this.ready_) --> showPlaceholder

Comment (dvoytenko) [reply extensions/amp-apester-media/0.1/amp-apester-media.js:164]: Hmm. I think this should simply be: before layout complete show placeholder, after layout complete don't show placeholder.

Comment (samouri) [reply extensions/amp-apester-media/0.1/amp-apester-media.js:164]: Reading into the historical context, and I'm still not sure why this component has custom placeholder info: https://github.com/ampproject/amphtml/pull/4291

Comment (dvoytenko) [reply extensions/amp-apester-media/0.1/amp-apester-media.js:164]: Not sure. But otherwise, the original rules seem pretty simple. Placeholder is hidden right after layout.
